executor/src/witgen/data_structures/identity.rs

```diff
@@ -0,0 +1,430 @@
+use core::fmt;
+use std::{collections::BTreeMap, ops::RangeFrom};
+
+use powdr_ast::{
+    analyzed::{
+        AlgebraicExpression, AlgebraicUnaryOperator, ConnectIdentity, Identity as AnalyzedIdentity,
+        LookupIdentity, PermutationIdentity, PhantomLookupIdentity, PhantomPermutationIdentity,
+        PolynomialIdentity, SelectedExpressions,
+    },
+    parsed::visitor::Children,
+};
+use powdr_number::FieldElement;
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct BusSend<T> {
+    /// The ID is globally unique among identities.
+    pub id: u64,
+    /// The ID of the bus interaction, used to find a matching receive.
+    pub interaction_id: AlgebraicExpression<T>,
+    /// The tuple sent to the bus, with a selector.
+    pub selected_tuple: SelectedExpressions<T>,
+}
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct BusReceive<T> {
+    /// The ID is globally unique among identities.
+    pub id: u64,
+    /// The ID of the bus interaction, to be matched with a send.
+    pub interaction_id: T,
+    /// The tuple sent to the bus, with a selector.
+    /// The selector is an expression that might or might not reference
+    /// the multiplicity column, but it should always evaluate to a
+    /// binary value. If it is zero, the multiplicity must be zero as well.
+    /// For example, it could also be binary fixed column, indicating
+    /// where the multiplicity can be non-zero.
+    pub selected_tuple: SelectedExpressions<T>,
+    /// The multiplicity with which the tuple is sent to the bus.
+    /// None if this interaction comes from the RHS of a native lookup,
+    /// as native lookup arguments don't require the multiplicity to be
+    /// committed.
+    /// Note that this is the absolute multiplicity, i.e., for bus receives,
+    /// the negation has been removed.
+    pub multiplicity: Option<AlgebraicExpression<T>>,
+}
+
+impl<T: FieldElement> BusSend<T> {
+    /// Tries to find the matching receive, which should succeed iff. the send has
```

Comment:
Why do you assume it is unique?

Reply:
You mean why the interaction ID uniquely identifies the receive? I think that would be always the case in practice. Otherwise the prover could choose where to route to, which would probably be a soundness issue, unless the two receivers compute the same function. I could add an assertion when building the `receives` map.

Reply:
At least conceptually, we should allow it (not saying we should support it practically). Just because there are multiple receivers listening for an ID doesn't mean they can all answer it conflict-free.

Reply:
BTW, I think the most generic algorithm is to try all machines (treating the interaction ID as just another parameter), and making sure exactly one machine succeeds. I think that would be a lot more complicated and potentially slower though.

Reply:
I'm proposing to change the documentation to say "a matching receive if it is unique."

Reply:
But that's clear from the type, no? I updated the docstring a bit, maybe you agree.